master/buildbot/db/logs.py

```diff
@@ -750,3 +759,71 @@ def _model_from_row(self, row):
             num_lines=row.num_lines,
             type=row.type,
         )
+
+
+async def _async_iter_on_pool(
+    reactor,
+    provider_threadpool,
+    /,
+    provider_fn: Callable[Concatenate[Callable[[_T], None], _P], None],
+    *args: _P.args,
+    **kwargs: _P.kwargs,
+) -> AsyncGenerator[_T, None]:
+    """
+    Utility to have a simple function returning an AsyncGenerator
+    when the workload yielding should be executed on a thread.
+    """
+    # create a single element queue as to
+    # occupy a thread of the pool
+    # avoiding too many compressed chunks in memory awaiting DB insert
+    # use 0 (unlimited) in tests as there isn't really a threadpool / reactor running
+    queue_maxsize = 1 if not get_is_in_unit_tests() else 0
```

Comment:
This function will be used for decompressing lines as well, right?

If yes, the queue_maxsize should come as an argument so that it would be possible to customize it on case by case basis. Higher queue size will be useful when the relative size of each item is small. In such case it is likely that performance would be significantly reducing due to threads blocking one another. The sender thread would repeatedly block on the queue being full and receiving thread would repeatedly block on the queue being empty. Only when sending and receiving threads are perfectly synchronized the amount of such blocks will be small.

Reply:
> This function will be used for decompressing lines as well, right?

yes, it's used in both compression (appendLog) and decompression (iter_log_lines, getLogLines) contexts.

> the queue_maxsize should come as an argument

I think it makes sense to provide flexibility for sure.

> The sender thread would repeatedly block on the queue being full and receiving thread would repeatedly block on the queue being empty. Only when sending and receiving threads are perfectly synchronized the amount of such blocks will be small.

At the moment, the thread that put in queue will block when it has done work and is ready to put in queue.
So as soon as consumer thread get the pending item from the queue, the producer thread will add the next item and go on to work on the next item.
Rationale here was to limit memory usage, at the cost of throughput.

A higher size would make sense to improve throughput, just not sure how to profile a valid use case to tweak it.

I used this setup to do some basic profiling when working on the fix: https://github.com/tdesveaux/buildbot-repro/tree/perf-regression-from-8157

(unrelated note, `RemoteCommand.split_line`, called from `RemoteCommand.remote_update` is blocking the mainthread in `LineBoundaryFinder.append`'s call to `re.Pattern.sub`, which should probably moved to a thread).

Reply:
> A higher size would make sense to improve throughput, just not sure how to profile a valid use case to tweak it.

I think it's reasonable to use gut feeling in this case. Log chunks are likely large, so in such case the queue size can be something like 10 (the extra memory size will still be less than 10 megabytes, so can still be ignored, as buildbot master is unlikely to have less than 1GB available). For something small like lines we can use larger number like 10000.

Reply:
https://github.com/buildbot/buildbot/pull/8166/commits/32ee89e35eda623bd6207f2b24265562e601404d

went with slightly different values, and commented on the rationale behind them.